### ISSUE TITLE
1.19.2 Endergetic Expansion Puffbug worldgen hang

### ISSUE THREAD

**Issue by Octelly** [MarkusBordihn/BOs-Adaptive-Performance-Tweaks#72]
**Describe the bug**
Minecraft worldgen hangs with the AP suite when trying to generate Endergetic Explansion Puffbug... errr, something? (not familiar with the content enough)

**To Reproduce**

1. Forge 1.19.2
2. Endergetic Expansion
3. AP Suite
4. Worldgen hangs, MC has to be killed from launcher

**Expected behavior**
Worldgen not hanging

**Is the issue happening in combination with other mods? (If yes, please provide a full list)**
Prism Launcher counts 527 mod files. I was benchmarking how far 32 GBs of RAM can get me after a recent upgrade. While I have not done thorough testing, worldgen resumed without any problems after disabling the AP suite. Note: I have checked mod pages and found that the Mods part is outdated and deprecated.

![image](https://github.com/user-attachments/assets/76c24210-a5d2-4e7b-88a3-d8c1496a9c90)

**Additional context**
This seems to be the relevant part of the log file
```
"Server thread" prio=5 Id=439 TIMED_WAITING on java.lang.String@6bd0965e
	at java.base@17.0.12/jdk.internal.misc.Unsafe.park(Native Method)
	-  waiting on java.lang.String@6bd0965e
	at java.base@17.0.12/java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:252)
	at TRANSFORMER/minecraft@1.19.2/net.minecraft.util.thread.BlockableEventLoop.m_5667_(BlockableEventLoop.java:521)
	at TRANSFORMER/minecraft@1.19.2/net.minecraft.util.thread.BlockableEventLoop.m_18701_(BlockableEventLoop.java:142)
	at TRANSFORMER/minecraft@1.19.2/net.minecraft.server.level.ServerChunkCache.m_7587_(ServerChunkCache.java:130)
	at TRANSFORMER/minecraft@1.19.2/net.minecraft.world.level.Level.m_6522_(MixinLevel.java:179)
	at TRANSFORMER/minecraft@1.19.2/net.minecraft.world.level.LevelReader.m_46819_(LevelReader.java:129)
	at TRANSFORMER/minecraft@1.19.2/net.minecraft.world.level.Level.m_6325_(MixinLevel.java:174)
	at TRANSFORMER/minecraft@1.19.2/net.minecraft.world.level.Level.m_46745_(MixinLevel.java:170)
	at TRANSFORMER/minecraft@1.19.2/net.minecraft.world.level.Level.m_7702_(MixinLevel.java:507)
	at TRANSFORMER/endergetic@4.0.0/com.teamabnormals.endergetic.common.entity.puffbug.PuffBug.getHive(PuffBug.java:515)
	at TRANSFORMER/endergetic@4.0.0/com.teamabnormals.endergetic.common.entity.puffbug.PuffBug.m_6785_(PuffBug.java:1105)
	at TRANSFORMER/adaptive_performance_tweaks_core@11.1.0/de.markusbordihn.adaptiveperformancetweakscore.entity.CoreEntityManager.isRelevantEntity(CoreEntityManager.java:480)
	at TRANSFORMER/adaptive_performance_tweaks_core@11.1.0/de.markusbordihn.adaptiveperformancetweakscore.entity.CoreEntityManager.handleEntityJoinLevelEvent(CoreEntityManager.java:125)
	at TRANSFORMER/adaptive_performance_tweaks_core@11.1.0/de.markusbordihn.adaptiveperformancetweakscore.entity.__CoreEntityManager_handleEntityJoinLevelEvent_EntityJoinLevelEvent.invoke(.dynamic)
	at MC-BOOTSTRAP/net.minecraftforge.eventbus/net.minecraftforge.eventbus.ASMEventHandler.invoke(ASMEventHandler.java:73)
	at MC-BOOTSTRAP/net.minecraftforge.eventbus/net.minecraftforge.eventbus.EventBus$$Lambda$5501/0x00007e9a94f96430.invoke(Unknown Source)
	at MC-BOOTSTRAP/net.minecraftforge.eventbus/net.minecraftforge.eventbus.EventBus.post(EventBus.java:315)
	at MC-BOOTSTRAP/net.minecraftforge.eventbus/net.minecraftforge.eventbus.EventBus.post(EventBus.java:296)
	at TRANSFORMER/minecraft@1.19.2/net.minecraft.world.level.entity.PersistentEntitySectionManager.m_157538_(PersistentEntitySectionManager.java:79)
	at TRANSFORMER/minecraft@1.19.2/net.minecraft.world.level.entity.PersistentEntitySectionManager.m_157604_(PersistentEntitySectionManager.java:121)
	at TRANSFORMER/minecraft@1.19.2/net.minecraft.world.level.entity.PersistentEntitySectionManager$$Lambda$103410/0x00007e9a94669c08.accept(Unknown Source)
	at TRANSFORMER/minecraft@1.19.2/net.minecraft.world.entity.EntityType$1.m_147060_(EntityType.java:529)
	at TRANSFORMER/minecraft@1.19.2/net.minecraft.world.entity.EntityType$1$$Lambda$96782/0x00007e9a9ccdf348.apply(Unknown Source)
	at java.base@17.0.12/java.util.Optional.map(Optional.java:260)
	at TRANSFORMER/minecraft@1.19.2/net.minecraft.world.entity.EntityType.m_20645_(EntityType.java:507)
	at TRANSFORMER/minecraft@1.19.2/net.minecraft.world.entity.EntityType$1.m_147056_(EntityType.java:528)
	at TRANSFORMER/minecraft@1.19.2/net.minecraft.world.entity.EntityType$1$$Lambda$96781/0x00007e9a9ccdf110.accept(Unknown Source)
	at java.base@17.0.12/java.util.ArrayList$ArrayListSpliterator.tryAdvance(ArrayList.java:1602)
	at TRANSFORMER/minecraft@1.19.2/net.minecraft.world.entity.EntityType$1.tryAdvance(EntityType.java:527)
	at java.base@17.0.12/java.util.Spliterator.forEachRemaining(Spliterator.java:332)
	at java.base@17.0.12/java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:762)
	at TRANSFORMER/minecraft@1.19.2/net.minecraft.world.level.entity.PersistentEntitySectionManager.m_157559_(PersistentEntitySectionManager.java:120)
	at TRANSFORMER/minecraft@1.19.2/net.minecraft.server.level.ServerLevel.m_143327_(ServerLevel.java:1517)
	at TRANSFORMER/minecraft@1.19.2/net.minecraft.server.level.ChunkMap.m_143064_(ChunkMap.java:666)
	at TRANSFORMER/minecraft@1.19.2/net.minecraft.server.level.ChunkMap.m_214898_(ChunkMap.java:683)
	at TRANSFORMER/minecraft@1.19.2/net.minecraft.server.level.ChunkMap$$Lambda$99669/0x00007e9a9d0e0240.m_196866_(Unknown Source)
	at TRANSFORMER/minecraft@1.19.2/net.minecraft.world.level.chunk.LevelChunk.m_62952_(LevelChunk.java:429)
	at TRANSFORMER/minecraft@1.19.2/net.minecraft.server.level.ChunkMap.m_214854_(ChunkMap.java:691)
	at TRANSFORMER/minecraft@1.19.2/net.minecraft.server.level.ChunkMap$$Lambda$96675/0x00007e9a9ccc3210.apply(Unknown Source)
	at MC-BOOTSTRAP/datafixerupper@5.0.28/com.mojang.datafixers.util.Either.lambda$mapLeft$0(Either.java:162)
	at MC-BOOTSTRAP/datafixerupper@5.0.28/com.mojang.datafixers.util.Either$$Lambda$96676/0x00007e9a9ccb89a0.apply(Unknown Source)
	at MC-BOOTSTRAP/datafixerupper@5.0.28/com.mojang.datafixers.util.Either$Left.map(Either.java:38)
	at MC-BOOTSTRAP/datafixerupper@5.0.28/com.mojang.datafixers.util.Either.mapLeft(Either.java:162)
	at TRANSFORMER/minecraft@1.19.2/net.minecraft.server.level.ChunkMap.m_214851_(ChunkMap.java:675)
	at TRANSFORMER/minecraft@1.19.2/net.minecraft.server.level.ChunkMap$$Lambda$96672/0x00007e9a9ccc2b78.apply(Unknown Source)
	at java.base@17.0.12/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:646)
	at java.base@17.0.12/java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:482)
	at TRANSFORMER/minecraft@1.19.2/net.minecraft.server.level.ChunkTaskPriorityQueueSorter.m_143188_(ChunkTaskPriorityQueueSorter.java:62)
	at TRANSFORMER/minecraft@1.19.2/net.minecraft.server.level.ChunkTaskPriorityQueueSorter$$Lambda$96660/0x00007e9a9ccc6390.run(Unknown Source)
	at TRANSFORMER/minecraft@1.19.2/net.minecraft.util.thread.BlockableEventLoop.m_6367_(BlockableEventLoop.java:157)
	at TRANSFORMER/minecraft@1.19.2/net.minecraft.server.level.ServerChunkCache$MainThreadExecutor.m_6367_(ServerChunkCache.java:535)
	at TRANSFORMER/minecraft@1.19.2/net.minecraft.util.thread.BlockableEventLoop.m_7245_(BlockableEventLoop.java:131)
	at TRANSFORMER/minecraft@1.19.2/net.minecraft.server.level.ServerChunkCache$MainThreadExecutor.m_7245_(ServerChunkCache.java:543)
	at TRANSFORMER/minecraft@1.19.2/net.minecraft.server.level.ServerChunkCache.m_8466_(ServerChunkCache.java:267)
	at TRANSFORMER/minecraft@1.19.2/net.minecraft.server.MinecraftServer.m_129961_(MinecraftServer.java:751)
	at TRANSFORMER/minecraft@1.19.2/net.minecraft.server.MinecraftServer.m_7245_(MinecraftServer.java:740)
	at TRANSFORMER/minecraft@1.19.2/net.minecraft.util.thread.BlockableEventLoop.m_18701_(BlockableEventLoop.java:140)
	at TRANSFORMER/minecraft@1.19.2/net.minecraft.server.MinecraftServer.m_130012_(MinecraftServer.java:726)
	at TRANSFORMER/minecraft@1.19.2/net.minecraft.server.MinecraftServer.m_130011_(MinecraftServer.java:658)
	at TRANSFORMER/minecraft@1.19.2/net.minecraft.server.MinecraftServer.m_206580_(MinecraftServer.java:244)
	at TRANSFORMER/minecraft@1.19.2/net.minecraft.server.MinecraftServer$$Lambda$95518/0x00007e9a9cb35ab0.run(Unknown Source)
	at java.base@17.0.12/java.lang.Thread.run(Thread.java:840)
```
I am only including this excerpt as it seems the most relevant and just launching Minecraft and closing it right after worldgen hanged (world loaded at the border of affected chunks) generated a log file of 8.4 MiB.

For full transparency: this happens on 64bit Temurin Java 17, Linux 6.10.10-zen1-1-zen with ``-XX:+UseG1GC -Dsun.rmi.dgc.server.gcInterval=2147483646 -XX:+UnlockExperimentalVMOptions -XX:G1NewSizePercent=20 -XX:G1ReservePercent=20 -XX:MaxGCPauseMillis=51 -XX:G1HeapRegionSize=32M -XX:+UseTransparentHugePages`` as the launch arguments.

**Comment by MarkusBordihn** [comment]
Thank you for the bug report. Unfortunately, I am unable to reproduce the issue. 
As you can see in the screenshot, the world is available and can be entered within a few seconds.

![image](https://github.com/user-attachments/assets/bee231bb-a5f5-4b0b-9564-aaeb49b2a7ee)

I'll take a closer look at the 'PuffBug' to see if I can identify any errors there. Does the same problem occur if the 'PuffBug' is added to the deny list? 
In this case, the entity check would be skipped, which seems to be causing the delay according to the log.

If this is an existing world, it’s possible that the PuffBug is a ticking entity, preventing the mod from performing the entity check.

**Comment by Octelly** [comment]
Thank you for the quick response!

I'll have to look more into it myself, but I'm a relatively fresh hire at a new position, so I haven't had that much time outside of weekends to troubleshoot this stuff.

Regarding the screenshot though (and sorry for not specifying), Puffbugs are only a thing in the End dimension. (it is **End**ergetic Expansion after all) The trace makes it look like the core module is at fault. Could that actually be the case or is that just how the code is architected?

**Comment by Enterprise12nx01** [comment]
i am having this issue too
https://pastebin.com/WE0jdCEh

**Comment by MarkusBordihn** [comment]
Thank you for sharing the details. It seems the issue occurs when the entity enters or leaves the world.

Technically, the related code only checks if the entity is a bee and whether it has a hive, as seen here:  
[CoreEntityManager.java#L481](https://github.com/MarkusBordihn/BOs-Adaptive-Performance-Tweaks/blob/40a748fe79e4e5a7a181b43d184a90b88dbd5056/core/src/main/java/de/markusbordihn/adaptiveperformancetweakscore/entity/CoreEntityManager.java#L481).

This suggests the problem is related either to 'The Endergetic Expansion' mod or another bee-related mod and how they handle the `hasHive` check. 
According to the logs you shared, it seems that calling `hasHive` triggers `getHive`, which attempts to access a chunk that is either not loaded or inaccessible. Normally, the bee's `hasHive` simply returns `true` or `false` without accessing any chunk data.

This points to an implementation issue within 'The Endergetic Expansion' or any bee-related mod that modifies this behavior, as seen here:  
[PuffBug.java#L515](https://github.com/team-abnormals/endergetic/blob/f49f630f5ecd3fb4ccd7a43de9347373b2a5087f/src/main/java/com/teamabnormals/endergetic/common/entity/puffbug/PuffBug.java#L515).

I was unable to reproduce the issue myself.  
![image](https://github.com/user-attachments/assets/b76073fd-b905-4e53-bff0-72fc1dbf880b)

Please report this issue to the 'The Endergetic Expansion' mod team for further investigation on their side.

**Comment by Octelly** [comment]
Thank you for your explanation, I have created an issue on their repo.

**Comment by Enterprise12nx01** [comment]
> Thank you for sharing the details. It seems the issue occurs when the entity enters or leaves the world.
> 
> Technically, the related code only checks if the entity is a bee and whether it has a hive, as seen here: [CoreEntityManager.java#L481](https://github.com/MarkusBordihn/BOs-Adaptive-Performance-Tweaks/blob/40a748fe79e4e5a7a181b43d184a90b88dbd5056/core/src/main/java/de/markusbordihn/adaptiveperformancetweakscore/entity/CoreEntityManager.java#L481).
> 
> This suggests the problem is related either to 'The Endergetic Expansion' mod or another bee-related mod and how they handle the `hasHive` check. According to the logs you shared, it seems that calling `hasHive` triggers `getHive`, which attempts to access a chunk that is either not loaded or inaccessible. Normally, the bee's `hasHive` simply returns `true` or `false` without accessing any chunk data.
> 
> This points to an implementation issue within 'The Endergetic Expansion' or any bee-related mod that modifies this behavior, as seen here: [PuffBug.java#L515](https://github.com/team-abnormals/endergetic/blob/f49f630f5ecd3fb4ccd7a43de9347373b2a5087f/src/main/java/com/teamabnormals/endergetic/common/entity/puffbug/PuffBug.java#L515).
> 
> I was unable to reproduce the issue myself. ![image](https://private-user-images.githubusercontent.com/12183293/374138529-b76073fd-b905-4e53-bff0-72fc1dbf880b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjgzNDM1NzEsIm5iZiI6MTcyODM0MzI3MSwicGF0aCI6Ii8xMjE4MzI5My8zNzQxMzg1MjktYjc2MDczZmQtYjkwNS00ZTUzLWJmZjAtNzJmYzFkYmY4ODBiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDA3VDIzMjExMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBkMDY3Yjk4ZWJhNTNkNzJhYjQ4OTNkMjEyNmUyZWQzMDQxOWNjNGEyYTdiNTE2ZjY2OWFlZTVmNjI0NzVkMzYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.V--XHIroAf674GS0ng_0D2d9VktqFDqqv3yhLSX0dUI)
> 
> Please report this issue to the 'The Endergetic Expansion' mod team for further investigation on their side.

Here is their response saying it's adaptive performance tweaks
https://github.com/team-abnormals/endergetic/issues/262#issuecomment-2397821414

**Comment by MarkusBordihn** [comment]
Submitted a PR to fix their code https://github.com/team-abnormals/endergetic/pull/264

**Comment by Enterprise12nx01** [comment]
> Submitted a PR to fix their code [team-abnormals/endergetic#264](https://github.com/team-abnormals/endergetic/pull/264)

A legend fixing someone else's mod! Thank you, that's above and beyond and I appreciate it. That team hasn't updated any of their mods in awhile so hopefully they push out the update or I can learn to package it myself

**Comment by Enterprise12nx01** [comment]
> Submitted a PR to fix their code [team-abnormals/endergetic#264](https://github.com/team-abnormals/endergetic/pull/264)

@MarkusBordihn they have seemed to unfortunately refused to fix it on their side, not sure if there's anything you can do on yours 😔?

**Comment by MarkusBordihn** [comment]
> > Submitted a PR to fix their code [team-abnormals/endergetic#264](https://github.com/team-abnormals/endergetic/pull/264)
> 
> @MarkusBordihn they have seemed to unfortunately refused to fix it on their side, not sure if there's anything you can do on yours 😔?

@Enterprise12nx01 
I've excluded their mod from this specific check, and the updated version should be live within the next few hours. 
However, please understand that if further issues with this mod arise, I won’t be able to provide patches on my side for a mod that is no longer being updated.

**Comment by Enterprise12nx01** [comment]
> > > Submitted a PR to fix their code [team-abnormals/endergetic#264](https://github.com/team-abnormals/endergetic/pull/264)
> > 
> > 
> > @MarkusBordihn they have seemed to unfortunately refused to fix it on their side, not sure if there's anything you can do on yours 😔?
> 
> @Enterprise12nx01 I've excluded their mod from this specific check, and the updated version should be live within the next few hours. However, please understand that if further issues with this mod arise, I won’t be able to provide patches on my side for a mod that is no longer being updated.

thank you!!
